### PR TITLE
land_detector initiate cycle immediately

### LINKED ISSUE #12190: land_detector complains about timeout
Recently with SITL I see this error output on startup:

```

ERROR [land_detector] Timed out while waiting for thread to start
INFO  [mavlink] mode: Normal, data rate: 4000000 B/s on udp port 14570 remote port 14550
INFO  [mavlink] mode: Onboard, data rate: 4000000 B/s on udp port 14580 remote port 14540
INFO  [mavlink] MAVLink only on localhost (set param MAV_BROADCAST = 1 to enable network)
...
```

To reproduce:
```
make px4_sitl jmavsim
```

@dagar any idea?

### PR DESCRIPTION
**Pull request by dagar**
- fixes #12190

### CODE REVIEW

**Comment by julianoes** [comment]
This did not work for me, same error message.

**Comment by dagar** [comment]
How about now? With `uORB::Subscription` we no longer have to worry about subscribing in the thread of execution, so some of the awkward startup dance goes away.

**Comment by julianoes** [comment]
Not sure what make of this :smile:: 

```
Thread 62 "px4" received signal SIGABRT, Aborted.
[Switching to Thread 0x7fffbaffd700 (LWP 32129)]
0x00007ffff7a81eb5 in raise () from /lib64/libc.so.6
Missing separate debuginfos, use: dnf debuginfo-install libgcc-9.1.1-1.fc30.x86_64 libstdc++-9.1.1-1.fc30.x86_64
(gdb) bt
#0  0x00007ffff7a81eb5 in raise () from /lib64/libc.so.6
#1  0x00007ffff7a6c895 in abort () from /lib64/libc.so.6
#2  0x00007ffff7e0f6da in ?? () from /lib64/libstdc++.so.6
#3  0x00007ffff7e1b61c in ?? () from /lib64/libstdc++.so.6
#4  0x00007ffff7e1b677 in std::terminate() () from /lib64/libstdc++.so.6
#5  0x00007ffff7e1c415 in __cxa_pure_virtual () from /lib64/libstdc++.so.6
#6  0x0000000000469137 in land_detector::LandDetector::_check_params (this=0x7fffac001a10, force=<optimized out>)
    at /home/julianoes/src/Firmware/src/modules/land_detector/LandDetector.cpp:156
#7  0x000000000046a2e9 in land_detector::MulticopterLandDetector::MulticopterLandDetector (this=0x7fffac001a10)
    at /home/julianoes/src/Firmware/src/modules/land_detector/MulticopterLandDetector.cpp:73
#8  0x0000000000468d0e in land_detector::LandDetector::task_spawn (argc=argc@entry=2, argv=argv@entry=0x7fffbaffcba8)
    at /home/julianoes/src/Firmware/src/modules/land_detector/land_detector_main.cpp:74
#9  0x0000000000419816 in ModuleBase<land_detector::LandDetector>::start_command_base (argv=0x7fffbaffcba8, argc=2)
    at /home/julianoes/src/Firmware/src/platforms/px4_module.h:317
#10 ModuleBase<land_detector::LandDetector>::main (argc=3, argv=0x7fffbaffcba0) at /home/julianoes/src/Firmware/src/platforms/px4_module.h:140
#11 0x00000000005af402 in px4_daemon::Pxh::process_line (line="land_detector start multicopter", silently_fail=silently_fail@entry=true)
    at /usr/include/c++/9/bits/basic_string.h:401
#12 0x00000000005af810 in px4_daemon::Pxh::process_line (line="land_detector start multicopter", silently_fail=silently_fail@entry=true)
    at /usr/include/c++/9/bits/basic_string.h:1031
#13 0x00000000005b07f2 in px4_daemon::Server::_handle_client (arg=0x7ffff0000b40) at /home/julianoes/src/Firmware/platforms/posix/src/px4_daemon/server.cpp:278
#14 0x00007ffff7f835a2 in start_thread () from /lib64/libpthread.so.0
#15 0x00007ffff7b45163 in clone () from /lib64/libc.so.6
```

File: src/modules/land_detector/LandDetector.cpp [hunk 1]
```diff
@@ -57,6 +57,16 @@ LandDetector::LandDetector() :
 	ScheduledWorkItem(px4::wq_configurations::hp_default),
 	_cycle_perf(perf_alloc(PC_ELAPSED, "land_detector_cycle"))
 {
+	_landDetected.timestamp = hrt_absolute_time();
+	_landDetected.freefall = false;
+	_landDetected.landed = true;
+	_landDetected.ground_contact = false;
+	_landDetected.maybe_landed = false;
```
Comment (mcsauder) [inline src/modules/land_detector/LandDetector.cpp]: Is it possible to default initialize these values at the struct definition instead?

File: src/modules/land_detector/land_detector_main.cpp [hunk 2]
```diff
@@ -81,28 +81,30 @@ int LandDetector::task_spawn(int argc, char *argv[])
...
+	_currentMode[sizeof(_currentMode) - 1] = '\0';
+
+	_object.store(obj);
+	_task_id = task_id_is_work_queue;
 	int ret = obj->start();
 
 	if (ret < 0) {
+		_object.store(nullptr);
+		_task_id = -1;
 		delete obj;
```
Comment (mcsauder) [inline src/modules/land_detector/land_detector_main.cpp]: I know it's not a change in this PR, but, do we need to add an `obj = nullptr` here?

**Comment by mcsauder** [comment]
Hi @dagar, I flight tested this PR and it's working quite right, it doesn't detect properly:
https://review.px4.io/plot_app?log=b5b3bc76-c1d3-445c-a29a-484a254a70ae

**Comment by dagar** [comment]
The problem was the `_check_params(true)` call in the constructor of LandDetector() (the base class). `LandDetector::_update_params()` is pure virtual.

**Pull request merged**